app/components/feedback-button/component.ts

```diff
@@ -0,0 +1,133 @@
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+import { inject as service } from '@ember/service';
+import config from 'ember-get-config';
+import $ from 'jquery';
+
+const {
+    microfeedback: { enabled, pageParams, url },
+    OSF: { backend },
+} = config;
+
+function sendFeedback(body: string, {
+    extra,
+    followup,
+    pageName,
+    userID,
+}: {
+    extra?: object,
+    followup: boolean,
+    pageName: string,
+    userID: string,
+}): Promise<any> {
+    const payload = {
+        body,
+        extra: {
+            followup,
+            user: userID ? `[${userID}|${window.location.origin}/${userID}/]` : undefined,
+            ...extra,
+        },
+    };
+
+    const params = [
+        (backend !== 'prod') ? { name: 'label', value: 'testing' } : undefined,
+        followup ? { name: 'label', value: 'followup' } : undefined,
+        ...Object.entries(pageParams[pageName] || {})
+            .map(([ name, value ]) => ({ name, value })),
+    ].filter((val) => !!val);
+
+    if (!url) {
+        return;
+    }
+
+    return $.ajax(`${url}?${$.param(params)}`, {
+        contentType: 'application/json',
+        data: JSON.stringify(payload),
+        dataType: 'json',
+        method: 'POST',
+    });
+}
+
+enum DialogState {
+    empty = 'empty',
+    active = 'active',
+    success = 'success',
+}
+
+export default class FeedbackButton extends Component.extend({
+    pageName:  null,
+    text: '',
+
+    actions: {
+        showDialog(this: FeedbackButton) {
+            this.set('state', DialogState.active);
+        },
+
+        hideDialog(this: FeedbackButton) {
+            this.set('state', DialogState.empty);
+            this.reset();
+        },
+
+        submit(this: FeedbackButton) {
+            const body = this.get('body');
+
+            // Dismiss if no input
+            if (!body) {
+                this.set('state', DialogState.empty);
+                this.reset();
+                return;
+            }
+
+            // Optimistically display success message
+            this.set('state', DialogState.success);
+            sendFeedback(
+                body,
+                {
+                    followup: this.get('followup'),
+                    pageName: this.get('pageName'),
+                    userID: this.get('currentUser.currentUserId'),
+                },
+            );
+            this.reset();
+        },
+    },
+}) {
+    private currentUser = service('currentUser');
```

Comment:
Initially, I tried to avoid using `Component.extend`, however there are a few problems with only using the class properties.
- The actions don't get merged with mixins, so we'll get errors with calling actions from the mixin. E.g. The analytics mixin click action errors.
- Handlebars attributes don't bind to the component properties. Basically it breaks data binding, which is a show-stopper. I mentioned this briefly in the scrum.

I agree and would love to use the latter syntax, however, I don't see a way forward on doing that at this time, without doing something equally atrocious.

In order to pass the recommended linting rules, marking a class property as public, private, or protected is required. Private seems like the most logical choice.

Reply:
Actions from mixins are a problem, that's annoying. Putting actions in `.extend({})` makes sense, then.

Handlebars attributes can bind to instance properties, you just have to be careful not to overwrite the initial bound value when declaring the property:
```
... extends Component {
    text: string;
}
```
instead of
```
... extends Component {
    text: string = '';
}
```
And I think that approach allows for typing those attributes in a much cleaner way than prototype properties.

Personally, I'd be in favor of disabling the `member-access` linting rule, since I don't think `private` everywhere gives us anything meaningful in components/controllers/routes \*shrug*

Reply:
Ya member-access isn't giving us much here. I'm with you on disabling it--at least for ember-specific things like components, routes, and controllers. I'll go ahead and make that change.